File: tools/profile-generator/extract/kbgen/catalog.py
"""The system catalog + section->system mapping table.

This is the data-driven core of extraction. To teach the generator a new system:
  1. add its key to CATALOG (mirrors docs/modules/_INDEX.md) with a default mode, and
  2. add one or more header aliases to SECTION_ALIASES.
Optionally register a rich extractor in extractors/ for structured signals (licenses,
groups, OU, ...). Everything else falls through to a presence signal or the unmodeled
report. Order in SECTION_ALIASES matters: most specific patterns first.
"""
from __future__ import annotations

import re

# system key -> default execution mode. Keys mirror docs/modules/_INDEX.md and the
# profile schema's known-keys list.
CATALOG: dict[str, str] = {
    "servicenow": "api",
    "m365": "api",
    "entra": "api",
    "exchange": "api",
    "active-directory": "api",
    "directory-sync": "api",
    "mimecast": "api",
    "proofpoint": "api",
    "adobe": "api",
    "google-workspace": "api",
    "knowbe4": "api",
    "spanning": "api",
    "sharepoint": "api",
    "zoom": "api",
    "slack": "api",
    "egnyte": "api",
    "egnyte-sync-server": "browser",
    "mdm": "api",
    "dropbox": "api",
    "perimeter81": "api",
    "teams": "api",
    "avd": "api",
    "1password": "api",
    "notion": "api",
    "tableau": "manual",
    "printix": "api",
    "address-book": "browser",
    "hardware": "manual",
    "workstation": "manual",
    "welcome-letter": "manual",
    "first-day-call": "manual",
    "archive": "api",
    "data-transfer": "api",
    "equipment-return": "manual",
    "case-resolution": "api",
}

# (regex pattern matched against the NORMALISED header, system key). First match wins,
# so order specific -> general. Normalisation (sectioning.normalize_header) lowercases,
# drops parentheticals/step-prefixes, and keeps the segment after an em/en dash.
SECTION_ALIASES: list[tuple[str, str]] = [
    # --- email security (specific before generic 'email') ---
    (r"\bmimecast\b", "mimecast"),
    (r"\bproofpoint\b", "proofpoint"),
    # --- Microsoft stack: order matters, exchange/entra before generic 365 ---
    (r"exchange( admin)?( center)?", "exchange"),
    (r"mailbox audit", "exchange"),
    (r"(microsoft |azure )?entra( id)?( admin)?", "entra"),
    (r"azure (active directory|ad|admin)", "entra"),
    # require a qualifier so a bare '365' (e.g. 'after 365 days') is NOT matched as m365.
    (r"(?:microsoft|ms|office) 365|365 admin(?: center)?|\bm365\b", "m365"),
    (r"\bo365\b", "m365"),
    (r"^admin center$", "m365"),
    (r"email account", "m365"),
    # --- on-prem identity ---
    (r"\bad sync\b", "directory-sync"),
    (r"directory sync", "directory-sync"),
    (r"active directory", "active-directory"),
    (r"domain( setup| part \d+)?$", "active-directory"),
    # --- core service desk ---
    (r"service ?now", "servicenow"),
    (r"^snow$", "servicenow"),
    (r"^sn$", "servicenow"),
    (r"case resolution|resolving (the )?case", "case-resolution"),
    # --- backup / storage ---
    (r"\bspanning\b", "spanning"),
    (r"\begnyte sync server\b", "egnyte-sync-server"),
    (r"\begnyte\b", "egnyte"),
    (r"\bdropbox\b", "dropbox"),
    (r"\bsharepoint\b", "sharepoint"),
    # --- identity-adjacent SaaS ---
    (r"g[\s-]?suite", "google-workspace"),
    (r"google( workspace)?", "google-workspace"),
    (r"knowbe4", "knowbe4"),
    (r"\bzoom\b", "zoom"),
    (r"\bslack\b", "slack"),
    (r"\bteams\b", "teams"),
    (r"adobe( acrobat)?( pro)?( ?/ ?creative cloud)?", "adobe"),
    (r"creative cloud", "adobe"),
    (r"\bnotion\b", "notion"),
    (r"\btableau\b", "tableau"),
    (r"1 ?password", "1password"),
    # --- network / endpoint ---
    (r"perimeter ?81", "perimeter81"),
    (r"vpn(/rds| setup)?", "perimeter81"),
    (r"\bavd\b|azure virtual desktop", "avd"),
    (r"\bjamf\b|\bintune\b|\baddigy\b|\bmdm\b", "mdm"),
    (r"\bprintix\b", "printix"),
    (r"printer address book|address book", "address-book"),
    # --- lifecycle / manual ---
    (r"welcome letter|info email", "welcome-letter"),
    (r"first[\s-]?day", "first-day-call"),
    (r"\bworkstation\b", "workstation"),
    (r"hardware(/data)?", "hardware"),
    (r"after \d+ days|^archive$", "archive"),
    (r"equipment return", "equipment-return"),
]

# Headers that are recognisably a system/vendor we do NOT model yet. Maps a pattern to a
# friendly guess label for the "systems detected but not yet modeled" report.
UNMODELED_GUESSES: list[tuple[str, str]] = [
    (r"\bduo\b", "Duo (MFA)"),
    (r"\bbox\b", "Box (storage)"),
    (r"salesforce", "Salesforce"),
    (r"idaptive", "Idaptive (CyberArk SSO)"),
    (r"lastpass", "LastPass"),
    (r"smartsheet", "Smartsheet"),
    (r"global relay", "Global Relay (archiving)"),
    (r"sso application", "SSO application (generic)"),
    (r"lob( applications| account removal)?", "LOB / line-of-business apps"),
]

# Procedural / structural headers that are NOT systems — ignored entirely (not reported
# as unmodeled, to keep that report signal-rich).
NOISE_HEADERS: set[str] = {
    "table of contents", "overview", "introduction", "summary", "notes",
    "final steps", "first steps", "prerequisites", "general", "background",
    "references", "user profile",
}


def classify_header(header: str) -> tuple[str, str | None]:
    """Classify a normalised header. Returns one of:
      ("system", <catalog key>)   — maps to a modelled system
      ("unmodeled", <guess|None>) — a system/vendor we don't model yet
      ("noise", None)             — procedural/structural, ignore
    """
    if header in NOISE_HEADERS:
        return ("noise", None)
    for pattern, key in SECTION_ALIASES:
        if re.search(pattern, header):
            return ("system", key)
    for pattern, guess in UNMODELED_GUESSES:
        if re.search(pattern, header):
            return ("unmodeled", guess)
    return ("unmodeled", None)

File: web/lib/modules/catalog.ts
// The module registry behind the Modules tab: every system the platform knows about, whether a
// runner EXECUTOR is built for it, and where its operator INSTRUCTIONS live (an in-app /help page).
//
// This is the one place to keep in sync when work ships:
//  - wire a new executor in the runner DISPATCH table  -> set executor: "built"
//  - ship an app/help/<slug>/page.tsx setup guide        -> set helpSlug
// The Modules page flags the gaps (built but no instructions; planned/no executor) automatically.

export type Executor = "built" | "manual" | "planned";
//  built   = a Coretelligent.* module wired into the runner DISPATCH table (or covered by m365/libs)
//  manual  = a human/checklist step by design — no API executor is expected
//  planned = in the catalog/profiles but no executor yet (the real backlog)

export type ModuleEntry = {
  key: string; // systemKey (or secret-scoped capability, e.g. exchange-onprem)
  name: string;
  group: "Core / identity" | "Email security" | "Apps & access" | "Security / endpoint" | "Notifications" | "Manual / hardware" | "Backlog (no executor)";
  executor: Executor;
  secret?: string; // logical secret name the runner brokers
  helpSlug?: string; // app/help/<slug> — the operator setup guide, if one exists
  note?: string;
};

export const MODULES: ModuleEntry[] = [
  // --- Core / identity ---
  { key: "servicenow", name: "ServiceNow", group: "Core / identity", executor: "built", secret: "servicenow", note: "case contact + work notes (runner lib)" },
  { key: "m365", name: "Microsoft 365", group: "Core / identity", executor: "built", secret: "m365-admin", helpSlug: "cloud-auth" },
  { key: "entra", name: "Entra", group: "Core / identity", executor: "built", secret: "m365-admin", helpSlug: "cloud-auth", note: "runs through the M365 module" },
  { key: "exchange", name: "Exchange Online", group: "Core / identity", executor: "built", secret: "m365-admin", helpSlug: "cloud-auth" },
  { key: "exchange-onprem", name: "Exchange (on-prem / hybrid)", group: "Core / identity", executor: "built", secret: "exchange-onprem", note: "hybrid mailbox session — needs a setup guide" },
  { key: "active-directory", name: "Active Directory", group: "Core / identity", executor: "built", secret: "ad-dc", note: "on-prem only — see docs/runner-dc-setup.md (needs an in-app guide)" },
  { key: "directory-sync", name: "Entra Connect sync", group: "Core / identity", executor: "built", secret: "ad-dc", note: "on-prem only (ADSync on the AAD Connect host)" },
  { key: "case-resolution", name: "Case resolution", group: "Core / identity", executor: "built" },

  // --- Email security ---
  { key: "mimecast", name: "Mimecast", group: "Email security", executor: "built", secret: "mimecast", helpSlug: "mimecast" },
  { key: "proofpoint", name: "Proofpoint", group: "Email security", executor: "planned", secret: "proofpoint" },
  { key: "spanning", name: "Spanning Backup", group: "Email security", executor: "built", secret: "spanning", helpSlug: "spanning" },

  // --- Apps & access ---
  { key: "google-workspace", name: "Google Workspace", group: "Apps & access", executor: "built", secret: "google-admin", helpSlug: "google" },
  { key: "adobe", name: "Adobe", group: "Apps & access", executor: "built", secret: "adobe", note: "spec at docs/modules/adobe.md — needs an in-app guide" },
  { key: "zoom", name: "Zoom", group: "Apps & access", executor: "built", secret: "zoom", helpSlug: "zoom" },
  { key: "knowbe4", name: "KnowBe4", group: "Apps & access", executor: "built", secret: "knowbe4", helpSlug: "knowbe4" },
  { key: "egnyte", name: "Egnyte", group: "Apps & access", executor: "built", secret: "egnyte", helpSlug: "egnyte" },
  { key: "perimeter81", name: "Perimeter 81", group: "Apps & access", executor: "built", secret: "perimeter81" },
  { key: "salesforce", name: "Salesforce", group: "Apps & access", executor: "built", secret: "salesforce", helpSlug: "salesforce" },
  { key: "hubspot", name: "HubSpot", group: "Apps & access", executor: "built", secret: "hubspot", helpSlug: "hubspot" },
  { key: "jira", name: "Jira", group: "Apps & access", executor: "built", secret: "jira", helpSlug: "jira" },
  { key: "sharepoint", name: "SharePoint", group: "Apps & access", executor: "planned", secret: "m365-admin" },
  { key: "slack", name: "Slack", group: "Apps & access", executor: "planned", secret: "slack" },
  { key: "teams", name: "Teams Phone", group: "Apps & access", executor: "planned", secret: "teams-admin" },
  { key: "dropbox", name: "Dropbox", group: "Apps & access", executor: "planned", secret: "dropbox" },
  { key: "1password", name: "1Password", group: "Apps & access", executor: "planned", secret: "1password" },
  { key: "notion", name: "Notion", group: "Apps & access", executor: "planned", secret: "notion" },
  { key: "printix", name: "Printix", group: "Apps & access", executor: "planned", secret: "printix" },
  { key: "avd", name: "Azure Virtual Desktop", group: "Apps & access", executor: "planned", secret: "m365-admin" },

  // --- Security / endpoint ---
  { key: "sentinelone", name: "SentinelOne", group: "Security / endpoint", executor: "built", secret: "sentinelone", helpSlug: "sentinelone" },
  { key: "duo", name: "Duo", group: "Security / endpoint", executor: "built", secret: "duo", helpSlug: "duo" },
  { key: "logicmonitor", name: "LogicMonitor", group: "Security / endpoint", executor: "built", secret: "logicmonitor", helpSlug: "logicmonitor" },
  { key: "mdm", name: "MDM (Addigy/Jamf/Intune)", group: "Security / endpoint", executor: "planned", secret: "mdm" },

  // --- Notifications ---
  { key: "notify", name: "Notify", group: "Notifications", executor: "built", note: "internal notifications" },
  { key: "xmatters", name: "xMatters", group: "Notifications", executor: "built", secret: "xmatters", helpSlug: "xmatters" },

  // --- Manual / hardware (no API executor by design) ---
  { key: "hardware", name: "Hardware", group: "Manual / hardware", executor: "manual" },
  { key: "workstation", name: "Workstation", group: "Manual / hardware", executor: "manual" },
  { key: "welcome-letter", name: "Welcome letter", group: "Manual / hardware", executor: "manual" },
  { key: "first-day-call", name: "First-day call", group: "Manual / hardware", executor: "manual" },
  { key: "equipment-return", name: "Equipment return", group: "Manual / hardware", executor: "manual" },
  { key: "address-book", name: "Printer address book", group: "Manual / hardware", executor: "manual", note: "browser fallback" },
  { key: "egnyte-sync-server", name: "Egnyte Sync Server", group: "Manual / hardware", executor: "manual", note: "browser fallback" },

  // --- Backlog: offboard helpers with no executor yet ---
  { key: "data-transfer", name: "Data transfer", group: "Backlog (no executor)", executor: "planned" },
  { key: "archive", name: "Archive (deferred)", group: "Backlog (no executor)", executor: "planned" },
];

export const helpHref = (m: ModuleEntry): string | null => (m.helpSlug ? `/help/${m.helpSlug}` : null);

// A "built" module with no operator guide is the actionable gap the Modules page highlights.
export const needsInstructions = (m: ModuleEntry): boolean => m.executor === "built" && !m.helpSlug;
